File: ml/congestion_estimator.py
"""
Cortexion ML — Congestion Estimator (Deliberately NOT ML)

With 2 demo vehicles you have nowhere near enough independent samples
per road segment to fit anything statistically meaningful. An online
exponentially-weighted estimator is the honest, correct tool for
low-volume, streaming, per-segment data.

A trained model here would be theater, not engineering, and a reviewer
who knows the field will notice.
"""

from collections import defaultdict
import time
import math
from typing import Optional


class SegmentSpeedEstimator:
    """
    EWMA per road segment for real-time congestion estimation.

    Half-life of 120 seconds means an observation's weight halves
    every 2 minutes — matching the timescale of urban traffic changes.
    """

    def __init__(self, half_life_s: float = 120.0):
        self.decay = math.log(2) / half_life_s
        self.state: dict[str, dict] = defaultdict(
            lambda: {'mean': None, 'last_update': None, 'sample_count': 0}
        )

    def update(self, segment_id: str, speed_kmh: float,
               now: Optional[float] = None) -> float:
        """
        Update a segment with a new speed observation.

        Returns the updated mean speed estimate.
        """
        now = now or time.time()
        s = self.state[segment_id]

        if s['mean'] is None:
            s['mean'] = speed_kmh
        else:
            dt = now - s['last_update']
            w = math.exp(-self.decay * dt)
            s['mean'] = w * s['mean'] + (1 - w) * speed_kmh

        s['last_update'] = now
        s['sample_count'] += 1
        return s['mean']

    def congestion_level(self, segment_id: str,
                          free_flow_kmh: float = 50.0) -> str:
        """
        Classify congestion level for a segment.

        Returns: 'clear', 'moderate', 'congested', or 'unknown'
        """
        s = self.state[segment_id]
        if s['mean'] is None:
            return 'unknown'

        ratio = s['mean'] / free_flow_kmh
        if ratio > 0.8:
            return 'clear'
        if ratio > 0.5:
            return 'moderate'
        return 'congested'

    def get_segment_info(self, segment_id: str,
                          free_flow_kmh: float = 50.0) -> dict:
        """Get detailed info for a segment."""
        s = self.state[segment_id]
        return {
            'segment_id': segment_id,
            'mean_speed': round(s['mean'], 1) if s['mean'] else None,
            'level': self.congestion_level(segment_id, free_flow_kmh),
            'sample_count': s['sample_count'],
            'last_update': s['last_update'],
        }

    def get_all_segments(self, free_flow_kmh: float = 50.0) -> list[dict]:
        """Get info for all tracked segments."""
        return [
            self.get_segment_info(seg_id, free_flow_kmh)
            for seg_id in self.state
            if self.state[seg_id]['mean'] is not None
        ]


def compute_segment_id(lat: float, lon: float) -> str:
    """
    Compute a segment ID from lat/lon coordinates.
    ~100m × 100m grid cells, matching firmware/common/segment.h.
    """
    lat_cell = int(lat * 1000)
    lon_cell = int(lon * 1000)
    return f"{lat_cell}:{lon_cell}"


if __name__ == '__main__':
    # Quick demo
    estimator = SegmentSpeedEstimator(half_life_s=60)
    seg = compute_segment_id(28.6139, 77.2090)

    # Simulate traffic slowing down
    for speed in [50, 45, 40, 30, 20, 15, 10]:
        mean = estimator.update(seg, speed)
        level = estimator.congestion_level(seg)
        print(f"Speed={speed:3d} → mean={mean:.1f} → {level}")

File: backend/src/congestion-estimator.js
/**
 * Cortexion Backend — Congestion Estimator
 *
 * Exponentially-weighted moving average (EWMA) per road segment.
 * DELIBERATELY NOT ML — with 2 demo vehicles, there are nowhere
 * near enough independent samples per road segment to fit anything.
 *
 * An EWMA that decays old observations handles the temporal aspect
 * (traffic changes over minutes) and is the standard tool for
 * streaming, low-volume estimation.
 */

const DEFAULT_HALF_LIFE_S = 120;  // 2 minutes
const DEFAULT_FREE_FLOW_KMH = 50; // Assumed free-flow speed for urban roads

export class CongestionEstimator {
  constructor(halfLifeS = DEFAULT_HALF_LIFE_S) {
    this.decay = Math.log(2) / halfLifeS;
    this.segments = new Map();
  }

  /**
   * Update a segment with a new speed observation.
   * @param {string} segmentId - Road segment identifier
   * @param {number} speedKmh - Observed speed
   * @param {number} [now] - Timestamp in seconds (defaults to Date.now()/1000)
   * @returns {number} Updated mean speed estimate
   */
  update(segmentId, speedKmh, now = Date.now() / 1000) {
    let seg = this.segments.get(segmentId);

    if (!seg) {
      seg = { mean: speedKmh, lastUpdate: now, sampleCount: 1 };
      this.segments.set(segmentId, seg);
      return speedKmh;
    }

    const dt = now - seg.lastUpdate;
    const w = Math.exp(-this.decay * dt);
    seg.mean = w * seg.mean + (1 - w) * speedKmh;
    seg.lastUpdate = now;
    seg.sampleCount++;

    return seg.mean;
  }

  /**
   * Get the congestion level for a segment.
   * @param {string} segmentId
   * @param {number} [freeFlowKmh] - Expected free-flow speed
   * @returns {{ level: string, ratio: number, meanSpeed: number, sampleCount: number }}
   */
  getLevel(segmentId, freeFlowKmh = DEFAULT_FREE_FLOW_KMH) {
    const seg = this.segments.get(segmentId);

    if (!seg) {
      return { level: 'unknown', ratio: null, meanSpeed: null, sampleCount: 0 };
    }

    const ratio = seg.mean / freeFlowKmh;
    let level;
    if (ratio > 0.8) level = 'clear';
    else if (ratio > 0.5) level = 'moderate';
    else level = 'congested';

    return {
      level,
      ratio: Math.round(ratio * 100) / 100,
      meanSpeed: Math.round(seg.mean * 10) / 10,
      sampleCount: seg.sampleCount,
    };
  }

  /**
   * Get all segments with their congestion levels.
   */
  getAllSegments(freeFlowKmh = DEFAULT_FREE_FLOW_KMH) {
    const result = {};
    for (const [id, _] of this.segments) {
      result[id] = this.getLevel(id, freeFlowKmh);
    }
    return result;
  }
}

/**
 * Compute a segment ID from lat/lon coordinates.
 * ~100m × 100m grid cells, matching firmware/common/segment.h.
 */
export function computeSegmentId(lat, lon) {
  const latCell = Math.floor(lat * 1000);
  const lonCell = Math.floor(lon * 1000);
  return `${latCell}:${lonCell}`;
}
